src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                Title = ServicesVSResources.Live_code_analysis,
+                ActionsAfterCompletion = CompletionActions.None
+            };
+        }
+
+        private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            // there is no concurrent call to this method since IDiagnosticService will serialize all
+            // events to preserve event ordering
+            var solution = e.Solution;
+            if (solution == null)
+            {
+                return;
+            }
+
+            var document = solution.GetDocument(e.DocumentId);
+            if (document != null)
+            {
+                ChangeProgress(string.Format(ServicesVSResources.Analyzing_0, document.Name ?? document.FilePath ?? "..."));
+                return;
+            }
+
+            var project = solution.GetProject(e.ProjectId);
+            if (project != null)
+            {
+                ChangeProgress(string.Format(ServicesVSResources.Analyzing_0, project.Name ?? project.FilePath ?? "..."));
+                return;
+            }
+        }
+
+        private void OnSolutionCrawlerProgressChanged(object sender, bool running)
+        {
+            // there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
+            // events to preserve event ordering
+            Started(running);
+        }
+
+        private void Started(bool running)
+        {
+            if (running)
+            {
+                // if there is any pending one. make sure it is finished.
+                _currentTask?.TrySetResult(true);
+
+                var taskHandler = _taskCenterService.PreRegister(_options, data: default);
+
+                _currentTask = new TaskCompletionSource<bool>();
+                taskHandler.RegisterTask(_currentTask.Task);
+
+                var data = new TaskProgressData
+                {
+                    ProgressText = null,
+                    CanBeCanceled = false,
+                    PercentComplete = null,
+                };
+
+                // report initial progress
+                taskHandler.Progress.Report(data);
+
+                // set handler
+                _taskHandler = taskHandler;
+            }
+            else
+            {
+                // stop progress
+                _currentTask?.TrySetResult(false);
+                _currentTask = null;
+
+                _taskHandler = null;
+            }
+        }
+
+        private void ChangeProgress(string message)
+        {
+            if (message == _lastMessage)
+            {
+                return;
+            }
+
+            var data = new TaskProgressData
+            {
+                ProgressText = message,
+                CanBeCanceled = false,
+                PercentComplete = null,
+            };
+
+            _lastMessage = message;
+            _taskHandler?.Progress.Report(data);
```

Comment:
💭 I am concerned about the performance impact of flooding the UI with update notifications, e.g. when analyzing thousands of small files with FSA enabled, or possibly when changing source control branches. However, I don't have enough evidence to say this will be a problem.

Reply:
sure

Reply:
It would be good to get explicit confirmation from @AArnott that the VS task center is ok with floods of notifications on the BG.  I think that's the case, but getting confirmation can't hurt :)